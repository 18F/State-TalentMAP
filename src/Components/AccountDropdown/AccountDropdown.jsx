import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import Dropdown, { DropdownTrigger, DropdownContent } from 'react-simple-dropdown';
import { EMPTY_FUNCTION, USER_PROFILE } from '../../Constants/PropTypes';
import { getAssetPath } from '../../utilities';

export class AccountDropdown extends Component {

  constructor(props) {
    super(props);
    this.hideDropdown = this.hideDropdown.bind(this);
    this.logout = this.logout.bind(this);
  }

  logout() {
    this.props.logoutRequest();
  }

  hideDropdown() {
    // Explicitly hide the dropdown using the built-in hide() function from react-simple-dropdown
    this.dropdown.hide();
  }

  render() {
    const userName = this.props.userProfile.user ? this.props.userProfile.user.username : '...';
    const avatar = getAssetPath('/assets/img/avatar.png');
    return (
      <Dropdown className="account-dropdown" ref={(dropdown) => { this.dropdown = dropdown; }} >
        <DropdownTrigger href="/#">
          <img
            alt="avatar"
            className="account-dropdown--avatar"
            src={avatar}
          />
          <span className="account-dropdown--name" id="account-username">{userName}</span>
        </DropdownTrigger>
        <DropdownContent>
          <div className="account-dropdown--identity account-dropdown--segment">
            <div>Signed in as</div>
            <strong>{userName}</strong>
          </div>
          <div className="account-dropdown--identity account-dropdown--segment">
            <Link to="/profile/dashboard" onClick={this.hideDropdown}>Profile</Link>
          </div>
          <div className="account-dropdown--identity account-dropdown--segment">
            <Link to="/login" onClick={this.logout}>Logout</Link>
          </div>
        </DropdownContent>
      </Dropdown>
    );
  }
}

AccountDropdown.propTypes = {
  logoutRequest: PropTypes.func,
  userProfile: USER_PROFILE,
};

AccountDropdown.defaultProps = {
  logoutRequest: EMPTY_FUNCTION,
  userProfile: {},
};

export default AccountDropdown;
