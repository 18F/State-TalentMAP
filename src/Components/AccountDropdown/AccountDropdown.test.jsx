import React from 'react';
import sinon from 'sinon';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import toJSON from 'enzyme-to-json';
import { AccountDropdown } from './AccountDropdown';

describe('AccountDropdown', () => {
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);

  it('is defined', () => {
    const accountDropdown = shallow(<AccountDropdown />);
    expect(accountDropdown).toBeDefined();
  });

  it('can take different props', () => {
    const accountDropdown = shallow(<AccountDropdown userProfile={{ user: { display_name: 'test' } }} />);
    expect(accountDropdown).toBeDefined();
  });

  it('can click the logout link', () => {
    const accountDropdown = shallow(<AccountDropdown />);

    // define the instance
    const instance = accountDropdown.instance();
    // spy the logout function
    const handleClickSpy = sinon.spy(instance, 'logout');
    // forceUpdate required for test to pass
    instance.forceUpdate();
    // click to logout
    accountDropdown.find('[href="/logout"]').simulate('click');
    // logout function should have been called once
    sinon.assert.calledOnce(handleClickSpy);
  });

  it('can call the hideDropdown function', () => {
    const accountDropdown = shallow(<AccountDropdown />);

    // define the instance
    const instance = accountDropdown.instance();
    instance.dropdown = { hide: () => {} };
    // spy the logout function
    const spy = sinon.spy(instance, 'hideDropdown');
    // click to logout
    instance.hideDropdown();
    // logout function should have been called once
    sinon.assert.calledOnce(spy);
  });

  it('does not display the name when shouldDisplayName is false', () => {
    const accountDropdown = shallow(<AccountDropdown shouldDisplayName={false} />);
    expect(accountDropdown.find('#account-username').exists()).toBe(false);
  });

  it('matches snapshot', () => {
    const accountDropdown = shallow(<AccountDropdown />);
    expect(toJSON(accountDropdown)).toMatchSnapshot();
  });

  it('matches snapshot when shouldDisplayName is true', () => {
    const accountDropdown = shallow(<AccountDropdown shouldDisplayName />);
    expect(toJSON(accountDropdown)).toMatchSnapshot();
  });

  it("can render the logged in user's name when shouldDisplayName is true", () => {
    const displayName = 'test';
    const accountDropdown = mount(<Provider store={mockStore({})}><MemoryRouter>
      <AccountDropdown shouldDisplayName userProfile={{ display_name: displayName }} />
    </MemoryRouter></Provider>);
    expect(accountDropdown.find('#account-username').text()).toBe(displayName);
  });
});
