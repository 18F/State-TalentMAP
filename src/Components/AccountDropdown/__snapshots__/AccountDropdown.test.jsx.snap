// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountDropdown matches snapshot 1`] = `
<Dropdown
  className="account-dropdown"
>
  <DropdownTrigger
    className=""
    href="/#"
  >
    <img
      alt="avatar"
      className="account-dropdown--avatar"
      src="/assets/img/avatar.png"
    />
    <span
      className="account-dropdown--name"
      id="account-username"
    >
      ...
    </span>
  </DropdownTrigger>
  <DropdownContent
    className=""
  >
    <div
      className="account-dropdown--identity account-dropdown--segment"
    >
      Signed in as 
      <strong>
        ...
      </strong>
    </div>
    <div
      className="account-dropdown--identity account-dropdown--segment"
    >
      <Link
        onClick={[Function]}
        replace={false}
        to="/profile"
      >
        Profile
      </Link>
    </div>
    <div
      className="account-dropdown--identity account-dropdown--segment"
    >
      <Link
        onClick={[Function]}
        replace={false}
        to="/login"
      >
        Logout
      </Link>
    </div>
  </DropdownContent>
</Dropdown>
`;
